<a href="https://colab.research.google.com/github/Janina-Ing/Analisis-y-visualizacion-de-datos/blob/main/Titanic_Alejandrina.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Titanic + Pandas Cheatsheet

# Configuración

In [20]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", 50)
pd.set_option("display.width", 120)

In [23]:
TITANIC_PATH = "Titanic.csv"
VERBOSE = True
def sep(title: str) -> None:
    """Imprime separadores para que la salida se vea profesional."""
    print("\n" + "=" * 80)
    print(title)
    print("=" * 80)


# Parte A ejercicio chaeatsheet pandas

# Exploración rápida

In [26]:
sep("A2) Exploración rápida")
print("\nhead(10):\n", df.head(10))
print("\ntail(5):\n", df.tail(5))
print("\nshape:", df.shape)
print("\ndtypes:\n", df.dtypes)
print("\ninfo():")
df.info()
print("\ndescribe():\n", df.describe())
print("\nNulos por columna:\n", df.isnull().sum())
print("\nValores únicos por columna:\n", df.nunique())
print("\nFrecuencia de valores (nombre):\n", df["nombre"].value_counts())
print("\nUso de memoria:\n", df.memory_usage(deep=True))


A2) Exploración rápida

head(10):
   nombre  edad  salario
0    Ana    28    50000
1   Luis    34    72000
2    Mia    22    41000

tail(5):
   nombre  edad  salario
0    Ana    28    50000
1   Luis    34    72000
2    Mia    22    41000

shape: (3, 3)

dtypes:
 nombre     object
edad        int64
salario     int64
dtype: object

info():
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   nombre   3 non-null      object
 1   edad     3 non-null      int64 
 2   salario  3 non-null      int64 
dtypes: int64(2), object(1)
memory usage: 204.0+ bytes

describe():
        edad       salario
count   3.0      3.000000
mean   28.0  54333.333333
std     6.0  15947.831619
min    22.0  41000.000000
25%    25.0  45500.000000
50%    28.0  50000.000000
75%    31.0  61000.000000
max    34.0  72000.000000

Nulos por columna:
 nombre     0
edad       0
salario    0
dtype: int6

# Selección y filtrado

In [28]:
sep("A3) Selección y filtrado")
print("\nColumna 'edad' (Serie):\n", df["edad"])
print("\nColumnas ['nombre','salario']:\n", df[["nombre", "salario"]])

print("\nFiltrar edad > 25:\n", df[df["edad"] > 25])
print("\nQuery edad > 25 and salario >= 50000:\n", df.query("edad > 25 and salario >= 50000"))
print("\nISIN nombre in ['Ana','Mia']:\n", df[df["nombre"].isin(["Ana", "Mia"])])

print("\nloc[0:2, 'nombre':'edad']:\n", df.loc[0:2, "nombre":"edad"])
print("\niloc[0:2, 0:2]:\n", df.iloc[0:2, 0:2])

print("\nfilter regex '^sal':\n", df.filter(regex="^sal"))


A3) Selección y filtrado

Columna 'edad' (Serie):
 0    28
1    34
2    22
Name: edad, dtype: int64

Columnas ['nombre','salario']:
   nombre  salario
0    Ana    50000
1   Luis    72000
2    Mia    41000

Filtrar edad > 25:
   nombre  edad  salario
0    Ana    28    50000
1   Luis    34    72000

Query edad > 25 and salario >= 50000:
   nombre  edad  salario
0    Ana    28    50000
1   Luis    34    72000

ISIN nombre in ['Ana','Mia']:
   nombre  edad  salario
0    Ana    28    50000
2    Mia    22    41000

loc[0:2, 'nombre':'edad']:
   nombre  edad
0    Ana    28
1   Luis    34
2    Mia    22

iloc[0:2, 0:2]:
   nombre  edad
0    Ana    28
1   Luis    34

filter regex '^sal':
    salario
0    50000
1    72000
2    41000


# Limpieza de datos

In [30]:
sep("A4) Limpieza (nulos, duplicados, rename, replace)")
df_dirty = df.copy()
df_dirty.loc[1, "salario"] = np.nan  # nulo
df_dirty = pd.concat([df_dirty, df_dirty.iloc[[0]]], ignore_index=True)  # duplicado
print("\nDF 'sucio':\n", df_dirty)

print("\nNulos por columna:\n", df_dirty.isnull().sum())


A4) Limpieza (nulos, duplicados, rename, replace)

DF 'sucio':
   nombre  edad  salario
0    Ana    28  50000.0
1   Luis    34      NaN
2    Mia    22  41000.0
3    Ana    28  50000.0

Nulos por columna:
 nombre     0
edad       0
salario    1
dtype: int64


# Fillna

In [32]:
df_filled = df_dirty.copy()
df_filled["salario"] = df_filled["salario"].fillna(0)
print("\nfillna(salario=0):\n", df_filled)


fillna(salario=0):
   nombre  edad  salario
0    Ana    28  50000.0
1   Luis    34      0.0
2    Mia    22  41000.0
3    Ana    28  50000.0


# drop_duplicates

In [34]:
df_no_dupes = df_filled.drop_duplicates(subset=["nombre", "edad", "salario"], keep="last")
print("\ndrop_duplicates:\n", df_no_dupes)


drop_duplicates:
   nombre  edad  salario
1   Luis    34      0.0
2    Mia    22  41000.0
3    Ana    28  50000.0


# Rename + replace

In [36]:
df_renamed = df_no_dupes.rename(columns={"salario": "salario_mxn"})
df_renamed["nombre"] = df_renamed["nombre"].replace({"Ana": "Ana (Edit)"})
print("\nrename + replace:\n", df_renamed)


rename + replace:
        nombre  edad  salario_mxn
1        Luis    34          0.0
2         Mia    22      41000.0
3  Ana (Edit)    28      50000.0


# A5 Groupby / agg/ pivot/ crosstab

In [38]:
sep("A5) Agrupación y agregación (groupby, agg, pivot_table, crosstab)")
df_grp = df.copy()
df_grp["depto"] = ["Ventas", "TI", "Ventas"]

print("\nPromedio salario por depto:\n", df_grp.groupby("depto")["salario"].mean())

agg = df_grp.groupby("depto").agg(
    salario_prom=("salario", "mean"),
    salario_max=("salario", "max"),
    empleados=("nombre", "count")
)
print("\nAgg múltiples:\n", agg)

pivot = pd.pivot_table(df_grp, values="edad", index="depto", aggfunc="mean")
print("\nPivot table (edad promedio por depto):\n", pivot)

ct = pd.crosstab(df_grp["depto"], df_grp["nombre"])
print("\nCrosstab depto vs nombre:\n", ct)


A5) Agrupación y agregación (groupby, agg, pivot_table, crosstab)

Promedio salario por depto:
 depto
TI        72000.0
Ventas    45500.0
Name: salario, dtype: float64

Agg múltiples:
         salario_prom  salario_max  empleados
depto                                       
TI           72000.0        72000          1
Ventas       45500.0        50000          2

Pivot table (edad promedio por depto):
         edad
depto       
TI      34.0
Ventas  25.0

Crosstab depto vs nombre:
 nombre  Ana  Luis  Mia
depto                 
TI        0     1    0
Ventas    1     0    1


# A6 Joins & mergé

In [40]:
sep("A6) Joins & merge (merge, concat, join)")
df1 = df[["nombre", "edad"]].copy()
df2 = df[["nombre", "salario"]].copy()

print("\nmerge inner on nombre:\n", pd.merge(df1, df2, on="nombre", how="inner"))

print("\nconcat vertical:\n", pd.concat([df1, df1], axis=0, ignore_index=True))

print("\nconcat horizontal:\n", pd.concat([df1.reset_index(drop=True), df2[["salario"]].reset_index(drop=True)], axis=1))

df1_i = df1.set_index("nombre")
df2_i = df2.set_index("nombre")
print("\njoin por índice:\n", df1_i.join(df2_i, how="inner"))


A6) Joins & merge (merge, concat, join)

merge inner on nombre:
   nombre  edad  salario
0    Ana    28    50000
1   Luis    34    72000
2    Mia    22    41000

concat vertical:
   nombre  edad
0    Ana    28
1   Luis    34
2    Mia    22
3    Ana    28
4   Luis    34
5    Mia    22

concat horizontal:
   nombre  edad  salario
0    Ana    28    50000
1   Luis    34    72000
2    Mia    22    41000

join por índice:
         edad  salario
nombre               
Ana       28    50000
Luis      34    72000
Mia       22    41000


# A7 Transformaciones

In [42]:
sep("A7) Transformaciones (apply, rolling, cumsum, pct_change, rank)")
df_trans = df.copy()
df_trans["nombre_upper"] = df_trans["nombre"].apply(lambda x: x.upper())
df_trans["rango_salario"] = df_trans["salario"].apply(lambda x: "alto" if x > 60000 else "bajo")
df_trans["rolling_avg_salario_2"] = df_trans["salario"].rolling(2).mean()
df_trans["cum_sum_salario"] = df_trans["salario"].cumsum()
df_trans["pct_change_salario"] = df_trans["salario"].pct_change()
df_trans["rank_salario"] = df_trans["salario"].rank(ascending=False)

print("\nDF transformado:\n", df_trans)


A7) Transformaciones (apply, rolling, cumsum, pct_change, rank)

DF transformado:
   nombre  edad  salario nombre_upper rango_salario  rolling_avg_salario_2  cum_sum_salario  pct_change_salario  \
0    Ana    28    50000          ANA          bajo                    NaN            50000                 NaN   
1   Luis    34    72000         LUIS          alto                61000.0           122000            0.440000   
2    Mia    22    41000          MIA          bajo                56500.0           163000           -0.430556   

   rank_salario  
0           2.0  
1           1.0  
2           3.0  


# Exportar

In [45]:
sep("A8) Exportar (to_csv)")
out_path = "salida_ejemplo.csv"
df_trans.to_csv(out_path, index=False)
print("Archivo generado:", out_path)


A8) Exportar (to_csv)
Archivo generado: salida_ejemplo.csv


# PARTE B Aplicar los ejercicios a Titanic

In [49]:
def cargar_titanic(path: str) -> pd.DataFrame:
    """Carga Titanic y hace ajustes básicos."""
    titanic = pd.read_csv(path)

    # Si viene con un índice guardado, lo quitamos
    if "Unnamed: 0" in titanic.columns:
        titanic = titanic.drop(columns=["Unnamed: 0"])

    return titanic


def parte_b_titanic(titanic: pd.DataFrame) -> pd.DataFrame:
    sep("PARTE B: Titanic (repetición de ejercicios con columnas reales)")

    # Validación básica de columnas esperadas
    required_cols = {"survived", "sex", "age", "fare", "embarked"}
    missing = required_cols - set(titanic.columns)
    if missing:
        print("\n[ADVERTENCIA] Faltan columnas esperadas:", missing)
        print("Revisa tu Titanic.csv y ajusta el script a tus columnas reales.\n")

    # B2) Exploración rápida
    sep("B2) Exploración rápida")
    print("shape:", titanic.shape)
    print("\nColumnas:\n", list(titanic.columns))

    print("\ninfo():")
    titanic.info()

    print("\ndescribe(include='all'):")
    # include='all' puede ser muy largo, pero es útil para entrega
    print(titanic.describe(include="all"))

    print("\nNulos por columna (top 10):")
    print(titanic.isnull().sum().sort_values(ascending=False).head(10))

    print("\nValores únicos (top 10):")
    print(titanic.nunique().sort_values(ascending=False).head(10))

    return titanic

# Validación básica de columnas esperadas

In [51]:
titanic_df = cargar_titanic(TITANIC_PATH)
titanic_df = parte_b_titanic(titanic_df)


PARTE B: Titanic (repetición de ejercicios con columnas reales)

B2) Exploración rápida
shape: (891, 15)

Columnas:
 ['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town', 'alive', 'alone']

info():
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    

# Selección y filtrado

In [56]:
sep("B3) Selección y filtrado")
if "age" in titanic_df.columns:
    print("\nSerie age (head):\n", titanic_df["age"].head())

subset_cols = [c for c in ["sex", "age", "fare"] if c in titanic_df.columns]
if subset_cols:
    print(f"\nSub-DataFrame {subset_cols} (head):\n", titanic_df[subset_cols].head())


B3) Selección y filtrado

Serie age (head):
 0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: age, dtype: float64

Sub-DataFrame ['sex', 'age', 'fare'] (head):
       sex   age     fare
0    male  22.0   7.2500
1  female  38.0  71.2833
2  female  26.0   7.9250
3  female  35.0  53.1000
4    male  35.0   8.0500


# Mayores de 50

In [58]:
if "age" in titanic_df.columns:
        print("\nPasajeros con age > 50 (head):\n", titanic_df[titanic_df["age"] > 50].head())


Pasajeros con age > 50 (head):
     survived  pclass     sex   age  sibsp  parch     fare embarked   class    who  adult_male deck  embark_town alive  \
6          0       1    male  54.0      0      0  51.8625        S   First    man        True    E  Southampton    no   
11         1       1  female  58.0      0      0  26.5500        S   First  woman       False    C  Southampton   yes   
15         1       2  female  55.0      0      0  16.0000        S  Second  woman       False  NaN  Southampton   yes   
33         0       2    male  66.0      0      0  10.5000        S  Second    man        True  NaN  Southampton    no   
54         0       1    male  65.0      0      1  61.9792        C   First    man        True    B    Cherbourg    no   

    alone  
6    True  
11   True  
15   True  
33   True  
54  False  


# Query

In [60]:
if {"sex", "pclass", "fare"}.issubset(titanic_df.columns):
        print("\nQuery: mujeres 1ra clase y fare>=50 (head):\n",
              titanic_df.query("sex == 'female' and pclass == 1 and fare >= 50").head())


Query: mujeres 1ra clase y fare>=50 (head):
     survived  pclass     sex   age  sibsp  parch      fare embarked  class    who  adult_male deck  embark_town alive  \
1          1       1  female  38.0      1      0   71.2833        C  First  woman       False    C    Cherbourg   yes   
3          1       1  female  35.0      1      0   53.1000        S  First  woman       False    C  Southampton   yes   
31         1       1  female   NaN      1      0  146.5208        C  First  woman       False    B    Cherbourg   yes   
52         1       1  female  49.0      1      0   76.7292        C  First  woman       False    D    Cherbourg   yes   
61         1       1  female  38.0      0      0   80.0000      NaN  First  woman       False    B          NaN   yes   

    alone  
1   False  
3   False  
31  False  
52  False  
61   True  


# Isin

In [62]:
if "embarked" in titanic_df.columns:
        print("\nEmbarked in ['S','C'] (head):\n", titanic_df[titanic_df["embarked"].isin(["S", "C"])].head())


Embarked in ['S','C'] (head):
    survived  pclass     sex   age  sibsp  parch     fare embarked  class    who  adult_male deck  embark_town alive  \
0         0       3    male  22.0      1      0   7.2500        S  Third    man        True  NaN  Southampton    no   
1         1       1  female  38.0      1      0  71.2833        C  First  woman       False    C    Cherbourg   yes   
2         1       3  female  26.0      0      0   7.9250        S  Third  woman       False  NaN  Southampton   yes   
3         1       1  female  35.0      1      0  53.1000        S  First  woman       False    C  Southampton   yes   
4         0       3    male  35.0      0      0   8.0500        S  Third    man        True  NaN  Southampton    no   

   alone  
0  False  
1  False  
2   True  
3  False  
4   True  


# Loc/iloc

In [64]:
print("\nloc[0:5]:\n", titanic_df.loc[0:5])
print("\niloc[0:5, 0:5]:\n", titanic_df.iloc[0:5, 0:5])


loc[0:5]:
    survived  pclass     sex   age  sibsp  parch     fare embarked  class    who  adult_male deck  embark_town alive  \
0         0       3    male  22.0      1      0   7.2500        S  Third    man        True  NaN  Southampton    no   
1         1       1  female  38.0      1      0  71.2833        C  First  woman       False    C    Cherbourg   yes   
2         1       3  female  26.0      0      0   7.9250        S  Third  woman       False  NaN  Southampton   yes   
3         1       1  female  35.0      1      0  53.1000        S  First  woman       False    C  Southampton   yes   
4         0       3    male  35.0      0      0   8.0500        S  Third    man        True  NaN  Southampton    no   
5         0       3    male   NaN      0      0   8.4583        Q  Third    man        True  NaN   Queenstown    no   

   alone  
0  False  
1  False  
2   True  
3  False  
4   True  
5   True  

iloc[0:5, 0:5]:
    survived  pclass     sex   age  sibsp
0         0       

# Regex filter

In [66]:
print("\nfilter regex 'embark':\n", titanic_df.filter(regex="embark").head())


filter regex 'embark':
   embarked  embark_town
0        S  Southampton
1        C    Cherbourg
2        S  Southampton
3        S  Southampton
4        S  Southampton


# Limpieza

In [68]:
sep("B4) Limpieza (nulos, types, duplicados, renombrar)")
print("\nNulos por columna (top 10):\n", titanic_df.isnull().sum().sort_values(ascending=False).head(10))


B4) Limpieza (nulos, types, duplicados, renombrar)

Nulos por columna (top 10):
 deck           688
age            177
embarked         2
embark_town      2
sex              0
pclass           0
survived         0
fare             0
parch            0
sibsp            0
dtype: int64


# Edad: imputar mediana

In [70]:
titanic_clean = titanic_df.copy()
if "age" in titanic_clean.columns:
    titanic_clean["age"] = titanic_clean["age"].fillna(titanic_clean["age"].median())

Embarkex : imputar moda

In [71]:
if "embarked" in titanic_clean.columns and titanic_clean["embarked"].isnull().any():
    titanic_clean["embarked"] = titanic_clean["embarked"].fillna(titanic_clean["embarked"].mode()[0])

# Convertir a category

In [72]:
 for c in ["sex", "embarked", "class", "who", "deck", "embark_town", "alive", "alone"]:
        if c in titanic_clean.columns:
            titanic_clean[c] = titanic_clean[c].astype("category")

# Duplicados

In [74]:
before = len(titanic_clean)
titanic_clean = titanic_clean.drop_duplicates()
after = len(titanic_clean)
print(f"\nDuplicados eliminados: {before - after}")


Duplicados eliminados: 110


# renombrar pclass

In [80]:
if "pclass" in titanic_clean.columns:
    titanic_clean = titanic_clean.rename(columns={"pclass": "pclass_num"})

print("\nHead titanic_clean:\n", titanic_clean.head())


Head titanic_clean:
    survived  pclass_num     sex   age  sibsp  parch     fare embarked  class    who  adult_male deck  embark_town  \
0         0           3    male  22.0      1      0   7.2500        S  Third    man        True  NaN  Southampton   
1         1           1  female  38.0      1      0  71.2833        C  First  woman       False    C    Cherbourg   
2         1           3  female  26.0      0      0   7.9250        S  Third  woman       False  NaN  Southampton   
3         1           1  female  35.0      1      0  53.1000        S  First  woman       False    C  Southampton   
4         0           3    male  35.0      0      0   8.0500        S  Third    man        True  NaN  Southampton   

  alive  alone  
0    no  False  
1   yes  False  
2   yes   True  
3   yes  False  
4    no   True  


# Agrupación y agregación

In [86]:
sep("B5) Agrupación y agregación")
if {"sex", "survived"}.issubset(titanic_clean.columns):
    survival_by_sex = titanic_clean.groupby("sex", observed=False)["survived"].mean().sort_values(ascending=False)
    print("\nTasa de supervivencia por sex:\n", survival_by_sex)

if {"pclass_num", "survived"}.issubset(titanic_clean.columns):
    agg_by_class = titanic_clean.groupby("pclass_num", observed=False).agg(
        pasajeros=("survived", "count"),
        tasa_supervivencia=("survived", "mean"),
        edad_prom=("age", "mean") if "age" in titanic_clean.columns else ("survived", "count"),
        fare_mediana=("fare", "median") if "fare" in titanic_clean.columns else ("survived", "mean")
    )
    print("\nAgregaciones por clase:\n", agg_by_class)

if {"sex", "class", "fare"}.issubset(titanic_clean.columns):
    pivot_fare = pd.pivot_table(
        titanic_clean,
        values="fare",
        index="sex",
        columns="class",
        aggfunc="mean",
        observed=False
    )
    print("\nPivot: fare promedio por sex y class:\n", pivot_fare)

if {"sex", "survived"}.issubset(titanic_clean.columns):
    ct = pd.crosstab(titanic_clean["sex"], titanic_clean["survived"])
    print("\nCrosstab sex vs survived:\n", ct)


B5) Agrupación y agregación

Tasa de supervivencia por sex:
 sex
female    0.740614
male      0.217213
Name: survived, dtype: float64

Agregaciones por clase:
             pasajeros  tasa_supervivencia  edad_prom  fare_mediana
pclass_num                                                        
1                 214            0.630841  36.913178       60.2875
2                 165            0.509091  29.829273       18.7500
3                 402            0.258706  25.657139        8.0500

Pivot: fare promedio por sex y class:
 class        First     Second      Third
sex                                     
female  106.521774  22.194921  15.704004
male     67.552617  21.550136  12.590801

Crosstab sex vs survived:
 survived    0    1
sex               
female     76  217
male      382  106


# Join & Merge

In [87]:
sep("B6) Joins & merge (didáctico)")
join_left_cols = [c for c in ["sex", "age", "pclass_num"] if c in titanic_clean.columns]
join_right_cols = [c for c in ["fare", "embarked"] if c in titanic_clean.columns]

left = titanic_clean[join_left_cols].copy().reset_index().rename(columns={"index": "passenger_id"})
right = titanic_clean[join_right_cols].copy().reset_index().rename(columns={"index": "passenger_id"})

merged = pd.merge(left, right, on="passenger_id", how="inner")
print("\nMerged head:\n", merged.head())

print("\nConcat vertical demo:\n", pd.concat([merged.head(2), merged.head(2)], axis=0, ignore_index=True))

left_i = left.set_index("passenger_id")
right_i = right.set_index("passenger_id")
print("\nJoin por índice head:\n", left_i.join(right_i, how="inner").head())


B6) Joins & merge (didáctico)

Merged head:
    passenger_id     sex   age  pclass_num     fare embarked
0             0    male  22.0           3   7.2500        S
1             1  female  38.0           1  71.2833        C
2             2  female  26.0           3   7.9250        S
3             3  female  35.0           1  53.1000        S
4             4    male  35.0           3   8.0500        S

Concat vertical demo:
    passenger_id     sex   age  pclass_num     fare embarked
0             0    male  22.0           3   7.2500        S
1             1  female  38.0           1  71.2833        C
2             0    male  22.0           3   7.2500        S
3             1  female  38.0           1  71.2833        C

Join por índice head:
                  sex   age  pclass_num     fare embarked
passenger_id                                            
0               male  22.0           3   7.2500        S
1             female  38.0           1  71.2833        C
2             fema

# Transformaciones/Features nuevas

In [90]:
sep("B7) Transformaciones / Features")
t = titanic_clean.copy()


B7) Transformaciones / Features


# Age of group

In [89]:
sep("B7) Transformaciones (aplicando ideas de A7 a Titanic)")

# Age of group
if "age" in titanic_clean.columns:
    bins = [0, 12, 18, 35, 60, 80]
    labels = ["child", "teen", "young adult", "adult", "senior"]
    titanic_clean["age_group"] = pd.cut(titanic_clean["age"], bins=bins, labels=labels, right=False)
    print("\nAge Group (head):\n", titanic_clean[["age", "age_group"]].head())


B7) Transformaciones (aplicando ideas de A7 a Titanic)

Age Group (head):
     age    age_group
0  22.0  young adult
1  38.0        adult
2  26.0  young adult
3  35.0        adult
4  35.0        adult


# Family size

In [91]:
 if {"sibsp", "parch"}.issubset(t.columns):
        t["family_size"] = t["sibsp"] + t["parch"] + 1
        if "fare" in t.columns:
            t["fare_per_person"] = t["fare"] / t["family_size"]

# Rank FAre

In [95]:
if "fare" in t.columns:
        t["fare_rank"] = t["fare"].rank(ascending=False)

# Rolling

In [96]:
if {"age", "fare"}.issubset(t.columns):
    t_sorted = t.sort_values("age").reset_index(drop=True)
    t_sorted["fare_rolling_avg_10"] = t_sorted["fare"].rolling(10).mean()
    print("\nRolling ejemplo (head 15):\n", t_sorted[["age", "fare", "fare_rolling_avg_10"]].head(15))

print("\nHead con features:\n", t.head())


Rolling ejemplo (head 15):
      age      fare  fare_rolling_avg_10
0   0.42    8.5167                  NaN
1   0.67   14.5000                  NaN
2   0.75   19.2583                  NaN
3   0.83   29.0000                  NaN
4   0.83   18.7500                  NaN
5   0.92  151.5500                  NaN
6   1.00   20.5750                  NaN
7   1.00   46.9000                  NaN
8   1.00   39.6875                  NaN
9   1.00   39.0000             38.77375
10  1.00   15.7417             39.49625
11  1.00   11.1333             39.15958
12  1.00   37.0042             40.93417
13  2.00   26.0000             40.63417
14  2.00   27.9000             41.54917

Head con features:
    survived  pclass_num     sex   age  sibsp  parch     fare embarked  class    who  adult_male deck  embark_town  \
0         0           3    male  22.0      1      0   7.2500        S  Third    man        True  NaN  Southampton   
1         1           1  female  38.0      1      0  71.2833        C  First

# Exportar

In [98]:
sep("B8) Exportar resultados")
t.to_csv("titanic_limpio_y_features.csv", index=False)
print("Archivo generado: titanic_limpio_y_features.csv")


B8) Exportar resultados
Archivo generado: titanic_limpio_y_features.csv


# Main

In [99]:
def main() -> None:
    # Parte A
    parte_a_cheatsheet()

    # Parte B (Titanic)
    sep("CARGA DE TITANIC")
    try:
        titanic = cargar_titanic(TITANIC_PATH)
    except FileNotFoundError:
        print(f"[ERROR] No se encontró el archivo '{TITANIC_PATH}'.")
        print("Solución:")
        print("- Coloca Titanic.csv en la misma carpeta que este script, o")
        print("- Cambia la variable TITANIC_PATH a la ruta correcta.")
        return

    parte_b_titanic(titanic)

    sep("FIN")
    print("Script terminado. Revisa los CSV generados para tu evidencia.")


if __name__ == "__main__":
    main()



PARTE A: Cheatsheet (DataFrame pequeño)

A1) Crear DataFrame
  nombre  edad  salario
0    Ana    28    50000
1   Luis    34    72000
2    Mia    22    41000

CARGA DE TITANIC

PARTE B: Titanic (repetición de ejercicios con columnas reales)

B2) Exploración rápida
shape: (891, 15)

Columnas:
 ['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town', 'alive', 'alone']

info():
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-

# Respuestas
 1. En el dataset Titanic identifiqué que las columnas con más valores faltantes fueron deck (688 nulos) y age (177 nulos). Decidí imputar age con la mediana (28) porque es una medida robusta y no se ve tan afectada por valores extremos, manteniendo una edad “típica” cuando faltan datos.

2. La tasa de supervivencia cambia notablemente según sex (mujeres ≈ 0.74 vs hombres ≈ 0.19) y también según pclass (1ra ≈ 0.63, 2da ≈ 0.47, 3ra ≈ 0.24). Esto sugiere que el acceso a recursos/ubicación en el barco y las prioridades de rescate (por ejemplo “mujeres primero”) pudieron influir en el resultado.

3. Al crear variables derivadas como family_size, fare_per_person y age_group observé que:

Familias pequeñas (2–4) tienden a tener mayor supervivencia (por ejemplo, tamaño 4 ≈ 0.72) que viajar solo (tamaño 1 ≈ 0.30) o familias muy grandes (8 o 11 ≈ 0.00).

En fare_per_person, el cuartil más alto tiene mayor supervivencia (≈ 0.61) que el cuartil más bajo (≈ 0.26).

Por age_group, los niños muestran mayor supervivencia (≈ 0.58) que los mayores (≈ 0.23).
Esto ayuda a explicar cómo la condición socioeconómica (tarifa), la estructura familiar y la edad se relacionan con la probabilidad de sobrevivir.

4. En Pandas, las funciones que más me facilitaron el análisis fueron groupby() y pivot_table(), ya que me permitieron resumir rápidamente patrones (promedios/tasas de supervivencia) por categorías como sexo, clase y grupos de edad, comparando resultados de forma clara.